### ISSUE TITLE
Discovery Not Seeing Generic Methods, just Generic Classes

### ISSUE THREAD

**Issue by gilmorem560** [xunit/xunit#1840]
I've got a project (can't share, proprietary, can try and create a repro if post isn't enough) implementing EF and the unit-of-work pattern, .NET Core C# stuff. To simplify and standardize our repository layer, I've created most of our post methodology in a generic repository base class, with generic tags and constraints placed on each method depending on the data type (as enforced by some interfaces in my Entities project).

I figured since we're using generic underlying methods, I might as well make the tests generic too. I followed the instructions in this blog post and they worked beautifully: http://blog.ploeh.dk/2011/05/09/GenericunittestingwithxUnit.net/

However, they only worked beautifully for one of my methods, since it has the least constraints. Trying to implement others I got compiler errors that I wasn't meeting the additional constraints. I figured, why not put my constraints on the method level instead of the class level, like I'm doing in my repository class. However, when I create the methods as generic with type-interface constraints, suddenly discovery can't find them.

As an example of what I'm trying to do:

Say I have a RepositoryBase class, with:
```csharp
    public class RepositoryBase {
        Add<TEntity>() where TEntity: class, IEntity, IInterface2, ..., new();
        Delete<TEntity>() where TEntity : class, IEntity, new();
    }
```
I then have a RepositoryBaseTest class, with in memory unit tests:
```csharp
    public class RepositoryBaseTest {
        [Fact]
        AddMemGen<TEntity>() where TEntity : class, IEntity, IInterface2, ..., new();
        [Fact]
        DeleteMemGen<TEntity>() where TEntity : class, IEntity, new();
    }
```
These tests aren't found in test explorer

Say instead my RepositoryBaseTest class is declared
```csharp
    public abstract class RepositoryBaseTest<TEntity> where TEntity : class, IEntity, new() {
        [Fact]
        AddMemGen();
        [Fact]
        DeleteMemGen();
    }
```
Then any specific test class where I inherit RepositoryBaseTest<TEntity> will automatically discover and run AddMemGen() and DeleteMemGen() successfully. Any specific test class where I inherit RepositoryBaseTest with the generic parameters on the methods instead seems to completely ignore the tests. Is there any way to have test discovery discover generic methods with the constraint on the method level instead of the class level? Often a single base can define methods with a wide array of different constraints to allow centralization of base functionality without necessarily having to extend each method to *every* data type passed in. It helps enforce standards and allows use of interfaces almost as "tags" to indicate which base methods can interact with each of our datatypes.

I've already tried a whole host of abstract, virtual, static, this that and the other modifiers on my base test class and methods, and I just have not gotten discovery to find them, only when I define the generic and constraints on the class level rather than method, which defeats the purpose of my RepositoryBase. Curiously enough, even though discovery can't find them, if I go into one of the tests in my actual test classes, and start typing, Intellisense finds the generic method and allows me to call it. The compiler allows me to compile, I can create a "wrapper" that then calls the generic function, so the functions are there and accessible, they're just not being seen as tests...

Thank you!

Edit: Pardon all the edits, I'm about as fluent in markdown as I am in Chinese....

**Comment by gilmorem560** [comment]
Update, just went ahead and built a little repro since I had some time
[Repro.Generic.Tests.zip](https://github.com/xunit/xunit/files/2520053/Repro.Generic.Tests.zip)

Oddly enough, the tests were discovered, so I will check in my other project to see why they aren't even being discovered. That being said, they're failing with a message that Fact tests cannot be run on generic methods. I'm inheriting the base class down into the specific test classes (Add prefixed for method-level constraints, no prefix for working, class level constraints)

Is it because the type is explicit in the inheritance of the generic class, but can't be determined from the inheritance of the generic methods?

**Comment by bradwilson** [comment]
The issue you're running into is that open generic methods (like your example `AddMemGen<TEntity>() where TEntity : class, IEntity, IInterface2, ..., new();`) need to get closed somehow. When you're calling this code by hand, you close it by providing the type when invoking:

```csharp
// This is legal
AddMemGen<MyClass>();

// This is not
AddMemGen();
```

There is no way for the compiler to know, in the second example, what the type is supposed to be; if the generic method used some method parameters of the given type, it could be determined automatically, but that's not the case here.

We have much the same problem with unit tests: if you have some piece of data that we can use to infer the data type, we will do so automatically. With your above example, where there is no available data from which to infer the type, we must fail (just like the compiler would fail in my second example).

In addition, we do not support generic constraints (the `where` part of the method declaration).

So, basically, what you're doing is not supported for two separate reasons. The rule is: test methods may only be openly generic _if and only if_ it is a theory which takes a parameter which can be used to determine the target type during execution (and generic constraints are never supported).

**Comment by gilmorem560** [comment]
Well in that case, is there any way to just say something along the lines of run this test for all entities which inherit the set of interfaces in the where clause of the base method being tested? I mean ideally the idea here is that we have a generic method and we write one single test against it that tests all data types rather than having to implement 30 different tests that test the exact same logic. Additionally, the creation of new objects would (like with the repository) lead to automatic implementation of tests against the new data since the only requirement that the method also work for the new data is that it implements the interfaces we're constraining on.

If that's not possible, that's fine, just wanted to verify there wasn't something I'm missing here, I want to make the test project as simple as the actual repository itself, but if I need to have the tests explicitly split out, then that's just what I need to do. Either way I do appreciate what you've got going here, excellent project.

**Comment by bradwilson** [comment]
>Well in that case, is there any way to just say something along the lines of run this test for all entities which inherit the set of interfaces in the where clause of the base method being tested?

You could generate that type list as data for a Theory, and then either (a) create instances of those classes as the theory data, which would then close the generic for you, or  (b) change the methods to take a `Type` instead of being generic.

**Comment by gilmorem560** [comment]
I decided to go with the latter, manipulating based on types. I added the most broad interface constraint to my base test class, inherited that out as the base for all my specific test classes. I then tossed in an if statement checking if the TEntity passed in from the class implements the necessary interfaces in the specific test methods. If not, I just pass back as "passing", if it does match, then I actually run the test itself.

The only downside is this populates the test under *every* type, regardless of whether the underlying method actually applies or not. On any where the method doesn't apply, they just report back as passing. Short of discovery only populating the test instances for valid interface types, this is just about what I needed to be able to do.

If I provide a predictable type (by instantiating the base class with a specific type), can I then have discovery only discover the test if that type matches certain parameters (i.e. in the test method declaration, is there anything you can think of with facts/theories where I can constrain on if(typeof(Interface).IsAssignableFrom(typeof(T))) or something to that effect). If not, that's fine, this is still leaps and bounds past where I thought I'd get to on this.

**Comment by gilmorem560** [comment]
Not sure if this needs to be closed or anything, but just wanted to let you know I've got a solution now going forward.

Just to help anyone that might be reading along, I decided to put a loose constraint on the generic class, requiring IEntity where IEntity is my global Entity interface.

I then check using if(typeof(Interface).IsAssignableFrom(typeof(T))) and then run my underlying test by casting the type to the interface of my choosing, should this if verify that they are compatible.

It's not the cleanest code, but it does seem to work every single time and in every scenario I hoped it would. This'll eliminate tons of duplicated code, and allows for scalability in that all we need to implement this are the interface handles assigned to the entity types. Then base test classes can be automatically implemented for new entity types simply by creating a new test class and inheriting the base class, closing the generic with the strong type.
